### ISSUE TITLE
[java] UnusedPrivateMethod: false positive with inner class instance method passed as a function reference parameter

### ISSUE THREAD

**Issue by dimas-b** [pmd/pmd#3292]
**Affects PMD Version:** 6.34.0

**Rule:** [UnusedPrivateMethod](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedprivatemethod)

**Description:**

A `private` method on an inner class used as function parameter to another method call is reported as "unused"

**Code Sample demonstrating the issue:**

```java
package com.example;

import java.util.function.BiFunction;

public class TestUnusedPrivateMethodWithFunctionReference {

  private final String separator;

  public TestUnusedPrivateMethodWithFunctionReference(String separator) {
    this.separator = separator;
  }

  public static void main(String[] args) {
    new TestUnusedPrivateMethodWithFunctionReference("|").test();
  }

  private void test() {
    System.out.println(makeValue(Holder::combine));
  }

  private String makeValue(BiFunction<Holder, Holder, Holder> combiner) {
    return combiner.apply(new Holder("t1"), new Holder("t2")).value;
  }

  private class Holder {

    private final String value;

    public Holder(String value) {
      this.value = value;
    }

    private Holder combine(Holder other) {
      return new Holder(value + separator + other.value);
    }
  }
}
```

PMD reports:
```
[...local path to...]/com/example/TestUnusedPrivateMethodWithFunctionReference.java:48:	UnusedPrivateMethod:	Avoid unused private methods such as 'combine(Holder)'.
```

**Expected outcome:**

No false positive "unused method" reported.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

CLI: 
```
./run.sh pmd -d [...local path to...]/src -R rulesets/java/quickstart.xml -f text
```

**Comment by pbludov** [comment]
This also affects static methods:

```java
package com.example;

public class TestUnusedPrivateMethodWithInnerClassStaticFunction {

  public static void main(String[] args) {
    Holder.from(0);
  }

  private static class Holder {
    private static Holder from(int value) {
        return null;
    }
  }
}
```
UnusedPrivateMethod:	Avoid unused private methods such as 'from(int)'.

**Comment by julius-d** [comment]
Hi!
In my project the update from pmd version 6.32.0 to a more up to date version fails
due to two false positives from the UnusedPrivateMethod rule.
One of the cases in my project is like the example in the comment from pbludov.
The other case is like this:

```java
package com.example;

import java.util.List;

public class Foo {

  public List<String> makeBar() {
    return List.of(1, 2, 3).stream().map(it -> doIt(it)).toList();
  }

  private String doIt(int i) {
    return (new BarProvider()).doIt(i);
  }

  private class BarProvider {

    private String doIt(Integer it) {
      return "bar " + it;
    }
  }
}

```

This code snippet is accepted in pmd version 6.32.0 but rejected in pmd version 6.33.0 to 6.36.0 with
```
PMD Failure: com.example.Foo$BarProvider:17 Rule:UnusedPrivateMethod Priority:3 Avoid unused private methods such as 'doIt(Integer)'..
```

Used java version : 16.0.1
Used maven version : 3.6.3_1

BR
Julius

**Comment by jsotuyod** [comment]
This is no longer reproducible under PMD 7.0.0
